Use memprof status field to distinguish minimal sampling

NickBarnes opened the pull request:
Statmemprof can be used at very low (or zero) frequency `lambda`, logically sampling allocation but in fact never taking a sample. We avoid doing any work for this case (e.g. we don't initialise the PRNGs). With this PR we test for this case using a new status value `CONFIG_STATUS_SAMPLING_MIN`, instead of the distinguished value -Inf for 1/log(1-lambda) (saved in a boxed float). This should (a) be very slightly faster to set the memprof trigger, especially on platforms with expensive float operations and (b) avoid dereferencing the boxed float to test against -Inf, which reduces the fragility of this code (see issue #14300 and the resolutions #14304, #14305, #14306).

NickBarnes committed:
Add CONFIG_STATUS_SAMPLING_MIN for the lambda ~=0 case.

gasche commented on runtime/memprof.c:
```diff
@@ -2223,14 +2225,18 @@ CAMLprim value caml_memprof_start(value lv, value szv, value tracker)
   /* Buggy implementations of caml_log1p could produce a
    * one_log1m_lambda which is positive infinity or NaN, which would
    * cause chaos in the RNG, so we check against this and set
    * one_log1m_lambda to negative infinity (which we can test for). We
```
the "(which we can test for)" part of the comment is not necessary anymore

gasche commented on runtime/memprof.c:
```diff
@@ -2223,14 +2225,18 @@ CAMLprim value caml_memprof_start(value lv, value szv, value tracker)
+    one_log1m_lambda = (-INFINITY);
+    status = Val_int(CONFIG_STATUS_SAMPLING_MIN);
+  } else {
+    status = Val_int(CONFIG_STATUS_SAMPLING);
```
I would expect that you also set CONFIG_STATUS_SAMPLING_MIN when `one_log1m_lambda` is computed to be `-INFINITY`, whereas this branch seems to use CONFIG_STATUS_SAMPLING in this case.

gasche commented on runtime/memprof.c:
```diff
@@ -2320,12 +2325,14 @@ CAMLprim value caml_memprof_discard(value config)
   case CONFIG_STATUS_STOPPED: /* correct case */
     break;
-  case CONFIG_STATUS_SAMPLING:
+  case CONFIG_STATUS_SAMPLING: /* fall-through */
```
The runtime uses an explicit `fallthrough` macro nowadays.

gasche reviewed (commented):
I think that this PR is very reasonable (see minor comments), but I'm not sure that it addresses my fundamental worry that we are accessing data on the heap in the middle of the GC critical section. This change would have avoided the race (without making sure that the floats are in the major heap, so the GC would only rely on the stuff that was already explicitly allocated in the major heap), but I was hoping to make the code more robust by avoiding reading the heaps from the minor GC completely, and if I understand correctly we haven't solved this problem here.

NickBarnes commented:
Yes, this isn't intended to fix that problem, and I will review #14306. IMO this is just nicer - not accessing the boxed float to determine this constant predicate - and I should have done it this way the first time. The boxed float is needed to generate the random numbers, and it is now only accessed for that purpose.

In practice the accesses to the memprof heap blocks (the profile and the boxed floats) should all be safe now, since #14304 (as they are all allocated on the major heap, so will not be moved or overwritten by the GC); the only mutable word is the status, which is only written and read atomically.

I've addressed your review comments.

gasche commented on Changes:
```diff
@@ -436,6 +436,11 @@ Working version
+- #14384: use the statmemprof status word, rather than a distinguished
+   value of a boxed float, to identify a "not really sampling"
+   profile.
+  (Nick Barnes, review by ??)
```
the ?? would need an update

gasche reviewed (approved):
I believe that the PR is correct. I am not sure that it makes the code simpler (it moves a condition from one place to another), but I trust @NickBarnes' judgment as the author that he prefers this new formulation.
